### PR TITLE
Minimal TypeScript example.

### PR DESCRIPTION
**Pull request by QuantumInformation**
The beginnings of typescript!

Handy for working with AWS sdk.

### CODE REVIEW

**Comment by QuantumInformation** [comment]
@pmuens What do you think mate? Maybe it would be clearer as a sibling to the node rest example? I do feel TypeScript has a benefit as node+AWS usage.

**Comment by pmuens** [comment]
Nice! Thanks for that @QuantumInformation 🎉 

🤔 I think this should be at the root of the repository. Maybe something like `aws-node-typescript-rest-api-with-dynamodb`. So the naming is `<provider>-<runtime>-<language>-<projectName>`.

What do you think @nikgraf @DavidWells ?

**Comment by DavidWells** [comment]
Yeah a typescript example deserves its own top level folder =)

**Comment by QuantumInformation** [comment]
Ok, eta weekend)

**Comment by QuantumInformation** [comment]
what about aws-node-rest-typescript-api-with-dynamodb so it sits closer together in the project:
image

**Comment by pmuens** [comment]
Thanks for the update @QuantumInformation  👍 

IMHO I would still take the name `aws-node-typescript-rest-api-with-dynamodb`. Here's why:

- This way all the typescript project will be near each other
- IMO the REST part belongs to API and not to TypeScript
- IMO The TypeScript part belongs to the runtime so it's <provider>-<runtime>-<optionalLanguage>-<projectName>

Just my two cents 💰

**Comment by QuantumInformation** [comment]
OK guys, its been moved..

**Comment by pmuens** [comment]
Great! Thanks @QuantumInformation 👍

**Comment by QuantumInformation** [comment]
Any plans to merge this?

File: aws-node-typescript-rest-api-with-dynamodb/serverless.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+functions:
+  create:
+    handler: todos/create.create
+    events:
+      - http:
+          path: todos
+          method: post
+          cors: true
+
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/serverless.yml]: This space can be removed

File: aws-node-typescript-rest-api-with-dynamodb/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  "exclude": [
+    "node_modules"
+  ],
+
+  "types": [
+    "node",
+    "aws-sdk"
+  ]
+}
+
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/tsconfig.json]: This extra space can be removed

File: aws-node-typescript-rest-api-with-dynamodb/create.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const uuid = require('uuid');
+import {DynamoDB} from 'aws-sdk'
+//const AWS = require('aws-sdk'); // eslint-disable-line import/no-extraneous-dependencies
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: This should be removed

File: aws-node-typescript-rest-api-with-dynamodb/create.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const uuid = require('uuid');
+import {DynamoDB} from 'aws-sdk'
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: `{DynamoDB}` --> `{ DynamoDB }`

File: aws-node-typescript-rest-api-with-dynamodb/create.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const uuid = require('uuid');
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: `import uuid from 'uuid'`

File: aws-node-typescript-rest-api-with-dynamodb/README.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+`npm install -g typescript`
+
+then
+
+`npm i`
+
+You can then run the compiler by running `tsc` in this directory. It will pull the settings from .tsconfig and extra @types
+from package.json. The output create.js file is what will be uploaded by serverless.
+
+#
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/README.md]: Is this headline missing? / can it be removed?

File: aws-node-typescript-rest-api-with-dynamodb/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#Introduction
+
+TypeScript (ts) offers type safety which is helpful when working with the AWS SDK, which comes with ts definitions (d.ts)
+
+#compiling
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/README.md]: Should be `# Compiling`

File: aws-node-typescript-rest-api-with-dynamodb/README.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#Introduction
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/README.md]: Should be `# Introduction`

File: aws-node-typescript-rest-api-with-dynamodb/serverless.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
+service: serverless-rest-api-with-dynamodb
+
+frameworkVersion: ">=1.1.0 <2.0.0"
+
+provider:
+  name: aws
+  runtime: nodejs4.3
```
Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/serverless.yml]: Runtime could be updated to `6.10`

**Comment by QuantumInformation** [comment]
I'll do this ASAP as I'm giving a big talk on it soon)

**Comment by pmuens** [comment]
> I'll do this ASAP as I'm giving a big talk on it soon)

Great! Thank you very much @QuantumInformation 👍

Let us know if you need any help since you might have some time-pressure due to the talk.

BTW. In the meantime Graphcool released a TypeScript plugin which plays nicely with Serverless: https://github.com/graphcool/serverless-plugin-typescript (in case you don't know it already).

File: aws-node-typescript-rest-api-with-dynamodb/create.ts (hunk 4 above)
Comment (QuantumInformation) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: Should we add an .editorconfig?

Comment (QuantumInformation) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: and lint rule: 
image,
mind you, would need to add tslint.

Comment (pmuens) [inline aws-node-typescript-rest-api-with-dynamodb/create.ts]: > Should we add an .editorconfig?

> and lint rule

Yes, that sounds like a good idea 👍

**Comment by pmuens** [comment]
Thank you very much for updating the PR @QuantumInformation 👍

Just some minor things (the build is currently failing) and the PR should be in a good state to be merged!

Do you want to add the `editorconfig` and `linting` rules as well?

**Comment by QuantumInformation** [comment]
Yes I'll add them in.

**Comment by QuantumInformation** [comment]
All done..

**Comment by pmuens** [comment]
Well, ok. Looks like there's another minor issue.

Could you please update the README so that the build passes 😬

Thanks in advance @QuantumInformation (we're almost there!).

**Comment by QuantumInformation** [comment]
I tried serverless deploy with this example, but get:

image

**Comment by pmuens** [comment]
Hey @QuantumInformation thanks for updating the PR again 👍

🤔 interesting... Could you update your Serverless version (we've just published a patch release which fixes a problem with the automatic dev dependency exclusion).

Just a quick question: So you're uploading the transpiled function code, right?

**Comment by QuantumInformation** [comment]
Looks good to me:

image
maybe it shouldn't include the ts file?

**Comment by QuantumInformation** [comment]
I'm using serverless 7.8:

image

**Comment by QuantumInformation** [comment]
I tried deleting the ts file, same error.

**Comment by darrenhgc** [comment]
If I'm not mistaken, the serverless.yml expects the create file to be in a todos directory

**Comment by QuantumInformation** [comment]
@pmuens hi can I connect via skype or screen hero about this in prep for the talk tomorrow?

**Comment by QuantumInformation** [comment]
@darrenhgc it is the case in this PR.

**Comment by QuantumInformation** [comment]
Adding `  region: eu-west-2` resulted in no error.

**Comment by QuantumInformation** [comment]
I tried to test the endpoint:

`curl -X POST https://6gyss4e9jh.execute-api.eu-west-2.amazonaws.com/dev/todos --data '{ "text": "Learn Serverless" }'`

but get this error

```
nable to import module 'todos/create': Error
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
END RequestId: 5478f9cc-7638-11e7-92b3-139fc9c0a56e
REPORT RequestId: 5478f9cc-7638-11e7-92b3-139fc9c0a56e	Duration: 9.66 ms	Billed Duration: 100 ms 	Memory Size: 1024 MBMax Memory Used: 19 MB

START RequestId: 6fd5385f-7638-11e7-ab5c-fba339c63863 Version: $LATEST
Unable to import module 'todos/create': Error
```

I've included the generated js

**Comment by QuantumInformation** [comment]
Hmm, the js doesn't pass lint.

**Comment by QuantumInformation** [comment]
@darrenhgc sorry your right! I'm getting muddled with all the dirs.

Next issue is the js that the tsc generates:

```js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const uuid_1 = require("uuid");
const aws_sdk_1 = require("aws-sdk");
const dynamoDb = new aws_sdk_1.DynamoDB.DocumentClient();
module.exports.create = (event, context, callback) => {
    const timestamp = new Date().getTime();
    const data = JSON.parse(event.body);
    if (typeof data.text !== 'string') {
        console.error('Validation Failed');
        callback(new Error('Couldn\'t create the todo item.'));
        return;
    }
    const params = {
        TableName: process.env.DYNAMODB_TABLE,
        Item: {
            id: uuid_1.default.v1(),
```

```
2017-07-31 23:08:51.021 (+01:00)	d51ad49d-763c-11e7-846c-ef19c264eee1	TypeError: Cannot read property 'v1' of undefined
    at module.exports.create (/var/task/todos/create.js:17:31)```

**Comment by QuantumInformation** [comment]
So the last thing I tried is

`curl -X POST https://6gyss4e9jh.execute-api.eu-west-2.amazonaws.com/dev/todos --data '{ "text": "Learn Serverless" }'`

but now I'm getting

```
START RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1 Version: $LATEST
2017-07-31 23:31:46.120 (+01:00)	090b9d20-7640-11e7-b677-cdbdc5a645f1	{ ResourceNotFoundException: Requested resource not found
    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:48:27)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:678:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:680:12)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
  message: 'Requested resource not found',
  code: 'ResourceNotFoundException',
  time: 2017-07-31T22:31:46.119Z,
  requestId: 'NIPIBGA54LSUNUKITHSQBIG9QNVV4KQNSO5AEMVJF66Q9ASUAAJG',
  statusCode: 400,
  retryable: false,
  retryDelay: 34.56349359159895 }
2017-07-31 23:31:46.121 (+01:00)	090b9d20-7640-11e7-b677-cdbdc5a645f1	{"errorMessage":"Couldn't create the todo item.","errorType":"Error","stackTrace":["Response.dynamoDb.put (/var/task/todos/create.js:29:22)","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:360:18)","Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:105:20)","Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:77:10)","Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:678:14)","Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)","AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)","/var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:680:12)"]}
END RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1
REPORT RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1	Duration: 97.77 ms	Billed Duration: 100 ms 	Memory Size: 1024 MBMax Memory Used: 38 MB
```

I'm off to sleep, maybe catchup tomorrow.

**Comment by QuantumInformation** [comment]
can't see the dynamo table after serverless deploy either

image

**Comment by pmuens** [comment]
> @pmuens hi can I connect via skype or screen hero about this in prep for the talk tomorrow?

Hey @QuantumInformation I'm sorry , I'm currently on vacation for two weeks and only have gotten Wifi which is quite spotty. 😞

However I looked into the code and here are some observations:

The handler for you're function is pointing at `todos/create.create` so AWS assumes that there's a `todos` directory with a `create.js` file.

Looking at the dir structure in https://github.com/serverless/examples/pull/93#issuecomment-318905362 shows that there's no `todos` directory but the `create.js` file is in the root. So the handler must be update to `create.create`.

Regarding the DynamoDB table: Can you look into the AWS console and see if another table with the same name is already there? The `DeletionPolicy: Retain` config means that the table won't be removed when you delete the service.

I'll follow this thread via E-Mail notifications and chime in again when I have internet access again. Feel free to comment if this resolves your problem!

**Comment by QuantumInformation** [comment]
Thx @pmuens, I've checked all my AWS regions and can't find any tables with the same name, I deleted the last one last night.  I also moved the create.js to /todos

**Comment by pmuens** [comment]
> Thx @pmuens, I've checked all my AWS regions and can't find any tables with the same name, I deleted the last one last night. I also moved the create.js to /todos

Alright. Can you share the exact error message you get right now when you deploy. Furthermore can you run `Serverless deploy --verbose` so that we can see the CloudFormation output during deployment?

Furthermore it looks like the `uuid` package is not included properly since the call to the `v1` function fails.

Thanks in advance!

**Comment by QuantumInformation** [comment]
I resolved it! I change the name to `service: serverless-typescript-rest-api-with-dynamodb` done a sls remove on the old one and pushed data!
thx for helping!

image

**Comment by pmuens** [comment]
> I resolved it!

@QuantumInformation great! Glad to hear that you were able to resolve it! Merging... :shipit:

**Pull request merged**